### ISSUE TITLE
torch.nn.functional.interpolate(mode='nearest'): backward(grad_output) should check grad_output.numel() <= std::numeric_limits<int32_t>::max()

### ISSUE THREAD

**Issue by CaesarKingW** [pytorch/pytorch#117011]
### 🐛 Describe the bug

```python3
import torch

n, c, hi, wi = 1025, 2, 256, 256
ho, wo = 1024, 1024
input = torch.randn(n, c, hi, wi, dtype=torch.float32).cuda()
input.requires_grad_(True)
output_size = [ho, wo]
output = torch.nn.functional.interpolate(input, size=output_size, mode ='nearest')
grad_output = torch.randn(n, c, ho, wo, dtype=torch.float32).cuda()
output.backward(grad_output)
grad_input = input.grad
print(grad_input.cpu())
```

Why is there no error message when the size of grad_output in upsample::nearest2d backward() exceeds INT32_MAX, but only the wrong result is output? The values of grad_input are all 0 instead of correctly calculating the gradient.

From the source code, you can see that only TORCH_CHECK (grad_input.numel() <= std::numeric_limits<int32_t>::max()) is performed  and grad_output is not checked, but the index of grad_output uses the int type.

https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cuda/UpSampleNearest2d.cu#L154
![image](https://github.com/pytorch/pytorch/assets/37433562/2f62b371-9d59-4233-887d-050037a1c63b)

https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cuda/UpSampleNearest2d.cu#L412
![image](https://github.com/pytorch/pytorch/assets/37433562/bdebf2f7-d2e7-43d4-9c3d-919f46d2ae0b)

### Versions

- pytorch 1.13 or 2.1

cc @ezyang @gchanan @zou3519 @kadeng @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by janeyx99** [comment]
hipri for silent correctness

**Comment by lezcano** [comment]
We would accept a PR adding the relevant `TORCH_CHECK`.
